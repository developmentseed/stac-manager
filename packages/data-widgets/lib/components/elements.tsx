import React from 'react';
import {
  Box,
  Flex,
  forwardRef,
  Heading,
  IconButton,
  Button,
  HeadingProps,
  IconButtonProps,
  FlexProps,
  Text
} from '@chakra-ui/react';
import {
  CollecticonTrashBin,
  CollecticonPlusSmall
} from '@devseed-ui/collecticons-chakra';

export const Fieldset = forwardRef<FlexProps, 'div'>((props, ref) => {
  return (
    <Flex
      ref={ref}
      as='fieldset'
      flexDirection='column'
      gap={8}
      p={4}
      bg='base.50a'
      borderRadius='md'
      {...props}
    />
  );
});

export const FieldsetHeader = forwardRef<FlexProps, 'div'>((props, ref) => {
  return <Flex ref={ref} justifyContent='space-between' gap={4} {...props} />;
});

export const FieldsetBody = forwardRef<FlexProps, 'div'>((props, ref) => {
  return <Flex ref={ref} flexDirection='column' gap={4} {...props} />;
});

export const FieldsetFooter = forwardRef<FlexProps, 'div'>((props, ref) => {
  return <Flex ref={ref} gap={4} {...props} />;
});

export const FieldLabel = forwardRef<HeadingProps, 'span'>((props, ref) => {
  return (
    <Heading
      ref={ref}
      as='span'
      size='sm'
      display='inline-flex'
      alignItems='center'
      gap={2}
      {...props}
      sx={{
        ...(props?.sx ?? {}),
        small: {
          borderRadius: 'sm',
          bg: 'base.400a',
          color: 'surface.500',
          px: '0.5rem',
          fontSize: 'xs'
        }
      }}
    />
  );
});

export const FieldIconBtn = forwardRef<IconButtonProps, 'button'>(
  (props, ref) => {
    return (
      <IconButton
        ref={ref}
        colorScheme='base'
        variant='soft-outline'
        size='xs'
        {...props}
      />
    );
  }
);

export const FieldsetDeleteBtn = forwardRef<IconButtonProps, 'button'>(
  (props, ref) => {
    return (
      <FieldIconBtn
        ref={ref}
        size='sm'
        icon={<CollecticonTrashBin />}
        {...props}
      />
    );
  }
);

interface ArrayFieldsetProps {
  label?: React.ReactNode;
  isRequired?: boolean;
  children: React.ReactNode;
  onRemove?: () => void;
  onAdd?: () => void;
  addDisabled?: boolean;
  removeDisabled?: boolean;
}

export function ArrayFieldset(props: ArrayFieldsetProps) {
  const {
    label,
    isRequired,
    children,
    onRemove,
    onAdd,
    addDisabled,
    removeDisabled
  } = props;

  return (
    <Fieldset className='widget--array'>
      {(label || onRemove) && (
        <FieldsetHeader>
          {label && (
            <Box>
              <FieldLabel>
                {label}
                {isRequired && (
                  <Text
                    as='span'
                    color='danger.500'
                    role='presentation'
                    aria-hidden='true'
                  >
                    *
                  </Text>
                )}
              </FieldLabel>
            </Box>
          )}
          {onRemove && (
            <Box>
              <FieldsetDeleteBtn
                onClick={onRemove}
                isDisabled={removeDisabled}
                aria-label='Remove item'
              />
            </Box>
          )}
        </FieldsetHeader>
      )}
      <FieldsetBody>{children}</FieldsetBody>
      {onAdd && (
        <FieldsetFooter>
          <Button
            colorScheme='base'
            size='sm'
            onClick={onAdd}
            aria-label='Add item'
            leftIcon={<CollecticonPlusSmall />}
            isDisabled={addDisabled}
          >
            Add another
          </Button>
        </FieldsetFooter>
      )}
    </Fieldset>
  );
}
