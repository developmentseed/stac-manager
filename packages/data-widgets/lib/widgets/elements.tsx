import React from 'react';
import {
  Box,
  Flex,
  forwardRef,
  Heading,
  IconButton,
  Button,
  HeadingProps,
  IconButtonProps,
  FlexProps
} from '@chakra-ui/react';
import {
  CollecticonTrashBin,
  CollecticonPlusSmall
} from '@devseed-ui/collecticons-chakra';

export const Fieldset = forwardRef<FlexProps, 'div'>((props, ref) => {
  return (
    <Flex
      ref={ref}
      as='fieldset'
      flexDirection='column'
      gap={8}
      p={4}
      bg='base.50a'
      borderRadius='md'
      {...props}
    />
  );
});

export const FieldsetHeader = forwardRef<FlexProps, 'div'>((props, ref) => {
  return <Flex ref={ref} justifyContent='space-between' gap={4} {...props} />;
});

export const FieldsetBody = forwardRef<FlexProps, 'div'>((props, ref) => {
  return <Flex ref={ref} flexDirection='column' gap={4} {...props} />;
});

export const FieldsetFooter = forwardRef<FlexProps, 'div'>((props, ref) => {
  return <Flex ref={ref} gap={4} {...props} />;
});

export const FieldLabel = forwardRef<HeadingProps, 'span'>((props, ref) => {
  return (
    <Heading
      ref={ref}
      as='span'
      size='sm'
      display='inline-flex'
      alignItems='center'
      gap={2}
      {...props}
      sx={{
        ...(props?.sx ?? {}),
        small: {
          borderRadius: 'sm',
          bg: 'base.400a',
          color: 'surface.500',
          px: '0.5rem',
          fontSize: 'xs'
        }
      }}
    />
  );
});

export const FieldsetDeleteBtn = forwardRef<IconButtonProps, 'button'>(
  (props, ref) => {
    return (
      <IconButton
        ref={ref}
        colorScheme='base'
        variant='soft-outline'
        size='sm'
        icon={<CollecticonTrashBin />}
        {...props}
      />
    );
  }
);

interface ArrayFieldsetProps {
  label: React.ReactNode;
  children: React.ReactNode;
  onRemove?: () => void;
  onAdd?: () => void;
  addDisabled?: boolean;
  removeDisabled?: boolean;
}

export function ArrayFieldset(props: ArrayFieldsetProps) {
  const { label, children, onRemove, onAdd, addDisabled, removeDisabled } =
    props;

  return (
    <Fieldset className='widget--array'>
      <FieldsetHeader>
        <Box>
          <FieldLabel>{label}</FieldLabel>
        </Box>
        {onRemove && (
          <Box>
            <FieldsetDeleteBtn
              onClick={onRemove}
              isDisabled={removeDisabled}
              aria-label='Remove item'
            />
          </Box>
        )}
      </FieldsetHeader>
      <FieldsetBody>{children}</FieldsetBody>
      {onAdd && (
        <FieldsetFooter>
          <Button
            colorScheme='base'
            size='sm'
            onClick={onAdd}
            aria-label='Add item'
            leftIcon={<CollecticonPlusSmall />}
            isDisabled={addDisabled}
          >
            Add another
          </Button>
        </FieldsetFooter>
      )}
    </Fieldset>
  );
}
