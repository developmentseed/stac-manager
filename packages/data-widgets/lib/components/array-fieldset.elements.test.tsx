import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';
import { ChakraProvider } from '@chakra-ui/react';
import { ArrayFieldset } from './elements';

// Custom renderer to add context or providers if needed
const renderWithProviders = (
  ui: React.ReactNode,
  { renderOptions = {} } = {}
) => {
  // You can wrap the component with any context providers here
  return render(ui, {
    wrapper: ({ children }) => <ChakraProvider>{children}</ChakraProvider>,
    ...renderOptions
  });
};

describe('ArrayFieldset', () => {
  it('renders the label and children correctly', () => {
    renderWithProviders(
      <ArrayFieldset label='Test Label' isRequired>
        <div>Child Content</div>
      </ArrayFieldset>
    );

    expect(screen.getByText('Test Label')).toBeInTheDocument();
    expect(screen.getByText('*')).toBeInTheDocument();
    expect(screen.getByText('Child Content')).toBeInTheDocument();
  });

  it('calls onRemove when the remove button is clicked', () => {
    const onRemove = jest.fn();
    renderWithProviders(
      <ArrayFieldset label='Test Label' onRemove={onRemove}>
        <div>Child Content</div>
      </ArrayFieldset>
    );

    const removeButton = screen.getByLabelText('Remove item');
    fireEvent.click(removeButton);

    expect(onRemove).toHaveBeenCalledTimes(1);
  });

  it('disables the remove button when removeDisabled is true', () => {
    renderWithProviders(
      <ArrayFieldset label='Test Label' onRemove={() => {}} removeDisabled>
        <div>Child Content</div>
      </ArrayFieldset>
    );

    const removeButton = screen.getByLabelText('Remove item');
    expect(removeButton).toBeDisabled();
  });

  it('calls onAdd when the add button is clicked', () => {
    const onAdd = jest.fn();
    renderWithProviders(
      <ArrayFieldset label='Test Label' onAdd={onAdd}>
        <div>Child Content</div>
      </ArrayFieldset>
    );

    const addButton = screen.getByLabelText('Add item');
    fireEvent.click(addButton);

    expect(onAdd).toHaveBeenCalledTimes(1);
  });

  it('disables the add button when addDisabled is true', () => {
    renderWithProviders(
      <ArrayFieldset label='Test Label' onAdd={() => {}} addDisabled>
        <div>Child Content</div>
      </ArrayFieldset>
    );

    const addButton = screen.getByLabelText('Add item');
    expect(addButton).toBeDisabled();
  });

  it('does not render the remove button when onRemove is not provided', () => {
    renderWithProviders(
      <ArrayFieldset label='Test Label'>
        <div>Child Content</div>
      </ArrayFieldset>
    );

    expect(screen.queryByLabelText('Remove item')).not.toBeInTheDocument();
  });

  it('does not render the add button when onAdd is not provided', () => {
    renderWithProviders(
      <ArrayFieldset label='Test Label'>
        <div>Child Content</div>
      </ArrayFieldset>
    );

    expect(screen.queryByLabelText('Add item')).not.toBeInTheDocument();
  });

  it('renders layout correctly', () => {
    const { rerender, container } = renderWithProviders(
      <ArrayFieldset>
        <div>Child Content</div>
      </ArrayFieldset>
    );
    const fieldset = container.querySelector('fieldset');
    expect(fieldset).toMatchSnapshot();

    rerender(
      <ArrayFieldset label='Test Label'>
        <div>Child Content</div>
      </ArrayFieldset>
    );
    expect(fieldset).toMatchSnapshot();

    rerender(
      <ArrayFieldset label='Test Label' onAdd={() => {}}>
        <div>Child Content</div>
      </ArrayFieldset>
    );
    expect(fieldset).toMatchSnapshot();

    rerender(
      <ArrayFieldset label='Test Label' onAdd={() => {}} onRemove={() => {}}>
        <div>Child Content</div>
      </ArrayFieldset>
    );
    expect(fieldset).toMatchSnapshot();

    rerender(
      <ArrayFieldset
        label='Test Label'
        onAdd={() => {}}
        onRemove={() => {}}
        isRequired
      >
        <div>Child Content</div>
      </ArrayFieldset>
    );
    expect(fieldset).toMatchSnapshot();
  });
});
