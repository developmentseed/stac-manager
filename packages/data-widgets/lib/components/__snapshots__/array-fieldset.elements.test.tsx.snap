// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArrayFieldset renders layout correctly 1`] = `
<fieldset
  class="widget--array css-g02mvc"
>
  <div
    class="css-1811skr"
  >
    <div>
      Child Content
    </div>
  </div>
</fieldset>
`;

exports[`ArrayFieldset renders layout correctly 2`] = `
<fieldset
  class="widget--array css-g02mvc"
>
  <div
    class="css-3fqpsw"
  >
    <div
      class="css-0"
    >
      <span
        class="chakra-heading css-1b4wzgc"
      >
        Test Label
      </span>
    </div>
  </div>
  <div
    class="css-1811skr"
  >
    <div>
      Child Content
    </div>
  </div>
</fieldset>
`;

exports[`ArrayFieldset renders layout correctly 3`] = `
<fieldset
  class="widget--array css-g02mvc"
>
  <div
    class="css-3fqpsw"
  >
    <div
      class="css-0"
    >
      <span
        class="chakra-heading css-1b4wzgc"
      >
        Test Label
      </span>
    </div>
  </div>
  <div
    class="css-1811skr"
  >
    <div>
      Child Content
    </div>
  </div>
  <div
    class="css-1azotaz"
  >
    <button
      aria-label="Add item"
      class="chakra-button css-1ktmsdb"
      type="button"
    >
      <span
        class="chakra-button__icon css-1wh2kri"
      >
        <svg
          aria-hidden="true"
          class="chakra-icon css-9a149e"
          focusable="false"
          role="img"
          viewBox="0 0 16 16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 0h16v16H0z"
            fill="none"
          />
          <path
            d="M14 7H9V2H7v5H2v2h5v5h2V9h5z"
          />
        </svg>
      </span>
      Add another
    </button>
  </div>
</fieldset>
`;

exports[`ArrayFieldset renders layout correctly 4`] = `
<fieldset
  class="widget--array css-g02mvc"
>
  <div
    class="css-3fqpsw"
  >
    <div
      class="css-0"
    >
      <span
        class="chakra-heading css-1b4wzgc"
      >
        Test Label
      </span>
    </div>
    <div
      class="css-0"
    >
      <button
        aria-label="Remove item"
        class="chakra-button css-12ozy5e"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="chakra-icon css-9a149e"
          focusable="false"
          role="img"
          viewBox="0 0 16 16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 0h16v16H0z"
            fill="none"
          />
          <path
            d="M11 5h2v8.5c0 .825-.675 1.5-1.5 1.5h-7c-.825 0-1.5-.675-1.5-1.5V5h2v8h2V5h2v8h2zM2 2h12v2H2zm4-2h4v1H6z"
          />
        </svg>
      </button>
    </div>
  </div>
  <div
    class="css-1811skr"
  >
    <div>
      Child Content
    </div>
  </div>
  <div
    class="css-1azotaz"
  >
    <button
      aria-label="Add item"
      class="chakra-button css-1ktmsdb"
      type="button"
    >
      <span
        class="chakra-button__icon css-1wh2kri"
      >
        <svg
          aria-hidden="true"
          class="chakra-icon css-9a149e"
          focusable="false"
          role="img"
          viewBox="0 0 16 16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 0h16v16H0z"
            fill="none"
          />
          <path
            d="M14 7H9V2H7v5H2v2h5v5h2V9h5z"
          />
        </svg>
      </span>
      Add another
    </button>
  </div>
</fieldset>
`;

exports[`ArrayFieldset renders layout correctly 5`] = `
<fieldset
  class="widget--array css-g02mvc"
>
  <div
    class="css-3fqpsw"
  >
    <div
      class="css-0"
    >
      <span
        class="chakra-heading css-1b4wzgc"
      >
        Test Label
        <span
          aria-hidden="true"
          class="chakra-text css-ai0irq"
          role="presentation"
        >
          *
        </span>
      </span>
    </div>
    <div
      class="css-0"
    >
      <button
        aria-label="Remove item"
        class="chakra-button css-12ozy5e"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="chakra-icon css-9a149e"
          focusable="false"
          role="img"
          viewBox="0 0 16 16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 0h16v16H0z"
            fill="none"
          />
          <path
            d="M11 5h2v8.5c0 .825-.675 1.5-1.5 1.5h-7c-.825 0-1.5-.675-1.5-1.5V5h2v8h2V5h2v8h2zM2 2h12v2H2zm4-2h4v1H6z"
          />
        </svg>
      </button>
    </div>
  </div>
  <div
    class="css-1811skr"
  >
    <div>
      Child Content
    </div>
  </div>
  <div
    class="css-1azotaz"
  >
    <button
      aria-label="Add item"
      class="chakra-button css-1ktmsdb"
      type="button"
    >
      <span
        class="chakra-button__icon css-1wh2kri"
      >
        <svg
          aria-hidden="true"
          class="chakra-icon css-9a149e"
          focusable="false"
          role="img"
          viewBox="0 0 16 16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 0h16v16H0z"
            fill="none"
          />
          <path
            d="M14 7H9V2H7v5H2v2h5v5h2V9h5z"
          />
        </svg>
      </span>
      Add another
    </button>
  </div>
</fieldset>
`;
